src/main/java/org/folio/rest/impl/PostGobiOrdersHelper.java

```diff
@@ -248,75 +250,78 @@ private CompletableFuture<Boolean> checkExistingOrder(CompositePurchaseOrder com
+        LOGGER.error("Error looking up for existing PO Line Number: {}", compositePurchaseOrder.getPoNumber(), t);
         return compositePurchaseOrder;
       });
 
   }
+
   public Void handleError(Throwable throwable) {
-    final javax.ws.rs.core.Response result;
+    Throwable cause = throwable.getCause();
+    LOGGER.error("Exception placing order", cause);
+    asyncResultHandler.handle(Future.succeededFuture(mapExceptionToResponse(cause != null ? cause : throwable)));
+    return null;
+  }
 
-    logger.error("Exception placing order", throwable.getCause());
-    GobiResponse response = new GobiResponse();
-    response.setError(new ResponseError());
+  private static Response mapExceptionToResponse(Throwable t) {
+
+    String message = t.getMessage();
 
-    final Throwable t = throwable.getCause();
     if (t instanceof HttpException) {
-      final int code = ((HttpException) t).getCode();
-      final String message = t.getMessage();
-
-      switch (code) {
-      case 400:
-        response.getError().setCode(CODE_BAD_REQUEST);
-        response.getError().setMessage(HelperUtils.truncate(t.getMessage(), 500));
-        result = respond400WithApplicationXml(GobiResponseWriter.getWriter().write(response));
-        break;
-      case 500:
-        result = respond500WithTextPlain(message);
-        break;
-      case 401:
-        result = respond401WithTextPlain(message);
-        break;
-      default:
-        result = respond500WithTextPlain(message);
+      switch (((HttpException) t).getCode()) {
+        case 400:
+          return respond400WithApplicationXml(writeGobiResponse(CODE_BAD_REQUEST, message));
+        case 401:
+          return respond401WithTextPlain(message);
       }
-    } else if (t instanceof GobiPurchaseOrderParserException) {
-      response.getError().setCode(CODE_INVALID_XML);
-      response.getError().setMessage(HelperUtils.truncate(t.getMessage(), 500));
-      result = respond400WithApplicationXml(GobiResponseWriter.getWriter().write(response));
-    } else {
-      result = respond500WithTextPlain(throwable.getMessage());
     }
 
-    asyncResultHandler.handle(Future.succeededFuture(result));
-    return null;
+    if (t instanceof GobiPurchaseOrderParserException) {
+      return respond400WithApplicationXml(writeGobiResponse(CODE_INVALID_XML, message));
+    }
+
+    if (t instanceof GobiException) {
+      String errorCode = ((GobiException) t).getErrorCode().getCode();
+      return respond500WithApplicationXml(writeGobiResponse(errorCode, message));
+    }
+
+    return respond500WithApplicationXml(message);
   }
+
+  private static BinaryOutStream writeGobiResponse(String errorCode, String message) {
+    GobiResponse response = new GobiResponse();
+    response.setError(new ResponseError());
+    response.getError().setCode(errorCode);
+    response.getError().setMessage(message == null ? null : message.substring(0, Math.min(message.length(), 500)));
```

Comment:
```
response.getError().setMessage(StringUtils.substring(message, 0, MAX_MESSAGE_LENGTH));
```
This code also handle null, but interesting why this limit of 500 symbols needed, for me better to remove it

Reply:
It's not the stacktrace to care about response size so removing it...